WIP: soporte para documentos electronicos

ovnicraft opened the pull request:
Implementado documento factura electronica

Commit:
agregado codigo inicial para documentos electronicos

Commit:
FIX

Commit:
Implementacion de generación de xml y firma

Agregado métodos para generación de código de acceso, código de
seguridad y generación de XML.

TODO: mejorar la implementación del código de seguridad, se puede usar
una secuencia del sistema ?.

SPLIT: metodo de generación de XML, firma, retorno de WS.

Commit:
Split

separados en varios métodos para la generación del xml.
información tributaria.
informacion de factura.

TODO: validación y firma.

Commit:
Metodos para elemento factura

Definido un metodo para cada bloque de elemento factura.
metodo para validacion.

TODO: separar la firma digital del XML

Commit:
Metodo de firma

creado un método de firma para cualquier implementacion.
en base al XML

Commit:
paths

paths como constantes

Commit:
Numeracion de facturas de venta

Implementado metodo de numeracion automática de facturas de venta.
Mantiene el numero en internal number si ya la valida.
CHECK: es correcto no pedir un nuevo numero ?

Commit:
renombrado modulo

Commit:
Refactor

Reestructuracion de directorios para crear un modulo xades

Commit:
Servicio SRI

creada clase de servicio del SRI para definir todas las constantes

Commit:
Metodos XML

son métodos xml de representacion de datos de factura y deben estar en
la clase factura del erp.

Commit:
XSD

movidos a direcotrio de schemas

Commit:
Validacion XML

Agregada clase para validación contra schema de factura

Commit:
modulo SRI

creado modulo SRI

Commit:
WIP: servicios XML

llamada a validar XML según XSD definido

Commit:
Clean code

simplificado la llamada de generación de XML y firma digital

Commit:
Merge branch 'einvoice' into fix-einvoice

Conflicts:
	l10n_ec_invoice_sequence/invoice.py

ovnicraft commented on l10n_ec_einvoice/einvoice.py:
```diff
@@ -0,0 +1,324 @@
+        del codigo de 8 digitos
+        """
+        arreglo = invoice.origin.split('/')
+        return (arreglo[1] + arreglo[2])[2:10]        
```
Es mejor utilizar una secuencia del sistema

ovnicraft commented on l10n_ec_einvoice/einvoice.py:
```diff
@@ -0,0 +1,324 @@
+        company = invoice.company_id
+        partner = invoice.partner_id
+        infoFactura = etree.Element('infoFactura')
+        etree.SubElement(infoFactura, 'fechaEmision').text = time.strftime('%d/%m/%Y', obj.date_invoice)
```
utilizar constante de formato de fecha del sistema

ovnicraft commented on l10n_ec_einvoice/einvoice.py:
```diff
@@ -0,0 +1,324 @@
+        descuento = '0.00'
+        for line in invoice.invoice_line:
+            if line.product_id.default_code == 'DESC':
+                descuento = '%.2f' % (line.price_unit*-1)
```
esto por ? @alcidesrivera

ovnicraft commented on l10n_ec_einvoice/einvoice.py:
```diff
@@ -0,0 +1,324 @@
+                etree.SubElement(totalImpuesto, 'valor').text = '%.2f' % (tax.tax_amount)
+                totalConImpuestos.append(totalImpuesto)
+        infoFactura.append(totalConImpuestos)
+        etree.SubElement(infoFactura, 'propina').text = '0.0000'
```
el sistema no lo implementa y el SRI lo deja como opcional, debe eliminarse

ovnicraft commented on l10n_ec_einvoice/einvoice.py:
```diff
@@ -0,0 +1,324 @@
+            for tax_line in line.invoice_line_tax_id:
+                if tax_line.tax_group in ['vat', 'vat0', 'ice', 'other']:
+                    base_amount = cur_obj.compute(cr, uid, invoice.currency_id.id, company_currency, line.price_subtotal * tax_line.base_sign, context={'date': invoice.date_invoice or time.strftime('%Y-%m-%d')}, round=False)
+                    tax_amount = cur_obj.compute(cr, uid, obj.currency_id.id, company_currency, tax_line.amount * tax_line.tax_sign, context={'date': invoice.date_invoice or time.strftime('%Y-%m-%d')}, round=False)
```
evitar el calculo en este metodo

Commit:
Correccion

Llamada a métodos de SRIService

Commit:
FIX

Llamada a metodo de servicio

Commit:
Elementos XML SRI V1.8

separado elemento descuento
agregado elementos requeridos en version 1.8

Commit:
Correccion

correccion de importación de clase

Commit:
Tipo de emision

Corregido tipo de emision, prograpado desde petición.
9 digitos para numero de factura.

Commit:
FIXES

Correcciones en métodos, definidos como metodos de clase.

Commit:
FIX

numeracion y validacion de facturas de venta

Commit:
FIX

correccin para generación de código de acceso.

Commit:
FIX

Corrección en llamada de metodos.
Pendient FIX formato FECHA
ERROR: calculo de lineas

Commit:
Merge branch '7.0' into fix-einvoice

Conflicts:
	l10n_ec_withdrawing/withdrawing.py

ovnicraft commented:
Por resolver https://github.com/odoo-ecuador/odoo-ecuador/issues/50

ovnicraft commented:
Seria bueno si alguien lo agrega:

Redefinir el metodo de calculo de total del objeto account_invoice_line, agregar el campo para base imponible + iva, incluirlo en la funcion existente https://github.com/odoo/odoo/blob/7.0/addons/account/account_invoice.py#L1388

Es necesario redefinir el campo para agregar el atributo multi
https://github.com/odoo/odoo/blob/7.0/addons/account/account_invoice.py#L1431

Alguna ayuda ? @miltonlab @alcidesrivera 

La implementacion puede ser en la herencia de la clase account_invoice_line en el archivo l10n_ec_withdrawing/withdrawing.py

miltonlab commented:
Yo pensé que ya existía
Te refieres a agregarlo a odoo o a odoo-ecuador?

2015-01-13 2:09 GMT-05:00 Cristian Salamea notifications@github.com:

> Seria bueno si alguien lo agrega:
> 
> Redefinir el metodo de calculo de total del objeto account_invoice_line,
> agregar el campo para base imponible + iva, incluirlo en la funcion
> existente
> https://github.com/odoo/odoo/blob/7.0/addons/account/account_invoice.py#L1388
> 
> Es necesario redefinir el campo para agregar el atributo multi
> 
> https://github.com/odoo/odoo/blob/7.0/addons/account/account_invoice.py#L1431
> 
> Alguna ayuda ? @miltonlab https://github.com/miltonlab @alcidesrivera
> https://github.com/alcidesrivera
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/odoo-ecuador/odoo-ecuador/pull/48#issuecomment-69703920
> .

## 

/\/\;/-----------------------------------------------------
Milton  Labanda  [miltonlab]
Distro:        Debian GNU/Linux 6.0 Squeeze
Blog:          http://1000tonlab.wordpress.com
jabber:        miltonlab@jabber.org milotnlab@jabber.org
"... Solamente la libertad que se somete a la Verdad conduce a la persona
humana a su  verdadero bien...".  Karol Wojtyla
(:\ Usa Software Legal, usa Software Libre /:)

ovnicraft commented:
@miltonlab lo que sucede es que @alcidesrivera ha identificado que para la facturacion electronica se requiere el total con iva por detalle, el sistema guarda el consolidado de impuestos en account.invoice.tax.

Como se quiere esto se debe agregar.

Lo confirmamos con la espec del SRI ?

miltonlab commented:
por detalle == total con iva por cada linea del detalle (o item de la
factura) ?

2015-01-14 12:08 GMT-05:00 Cristian Salamea notifications@github.com:

> @miltonlab https://github.com/miltonlab lo que sucede es que
> @alcidesrivera https://github.com/alcidesrivera ha identificado que
> para la facturacion electronica se requiere el total con iva por detalle,
> el sistema guarda el consolidado de impuestos en account.invoice.tax.
> 
> Como se quiere esto se debe agregar.
> 
> Lo confirmamos con la espec del SRI ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/odoo-ecuador/odoo-ecuador/pull/48#issuecomment-69951425
> .

## 

/\/\;/-----------------------------------------------------
Milton  Labanda  [miltonlab]
Distro:        Debian GNU/Linux 6.0 Squeeze
Blog:          http://1000tonlab.wordpress.com
jabber:        miltonlab@jabber.org milotnlab@jabber.org
"... Solamente la libertad que se somete a la Verdad conduce a la persona
humana a su  verdadero bien...".  Karol Wojtyla
(:\ Usa Software Legal, usa Software Libre /:)

ovnicraft commented:
Linea del detalle o itema de factura.

No conozco el requerimiento especifico del SRI en el XML, si lo podemos confirmar mejor.

ovnicraft commented on l10n_ec_einvoice/einvoice.py:
```diff
@@ -0,0 +1,338 @@
+                    etree.SubElement(impuesto, 'baseImponible').text = '%.2f' % (base_amount)
+                    etree.SubElement(impuesto, 'valor').text = '%.2f' % (base_amount * tax_amount)
+                    impuestos.append(impuesto)
+            detalle.append(impuestos)
```
@miltonlab es para evitar el calculo en este for y solo leer los valores de los campos.
@alcidesrivera es importante detalle que campos se requieren exactamente (con retencion, con iva, etc) para implementarlos correctamente.

Commit:
Agregado campo

Calculo de campo con iva agregado en linea de factura.
Correcion en validación de datos cuando se duplica registro. L863

El cambio agrega un campo en el objeto, el campo es de tipo funcion por
lo que el sistema cuando actualice calculara el campo automáticamente
:-)

ovnicraft commented on l10n_ec_withdrawing/withdrawing.py:
```diff
@@ -1130,7 +1130,41 @@ def product_id_change(self, cr, uid, ids, product, uom_id, qty=0, name='', type=
+                res[line.id]['price_subtotal'] = cur_obj.round(cr, uid, cur, res[line.id]['price_subtotal'])
+                res[line.id]['amount_tax'] = cur_obj.round(cr, uid, cur, res[line.id]['amount_tax'])                
+        return res    
+
```
@alcidesrivera agregado un commit para el campo del iva en el objeto de linea de factura test :-)

codeadict commented on l10n_ec_einvoice/xades/xades.py:
```diff
@@ -0,0 +1,106 @@
+        """
+        Metodo que aplica la firma digital al XML
+        """
+        OPT_PATH = '/opt/facturas/'
```
por que **/opt/facturas/** no mejor tener un campo de configuración donde se defina el directorio aunque x default sea este, tal vez pensar en integración con Google Drive y Dropbox a futuro para guardar los XML´s

ovnicraft replied:
@miltonlab @codeadict @alcidesrivera 
El metodo apply_digital_signature, necesita una depuracion exaustiva.
Existen varios requerimientos aca:
- firmar el XML: como firmar ? java es la unica solucion ? de serlo nos toca implementar un backend usable, luego irlo mejorando y salta una duda, que priorizamos ? poder usarlo o que haya calidad sin importar tiempo ?
- Donde almacenar el certificado
- zipear el xml firmado con el checksum
- generar el pdf en RIDE

Considero que todo el zip debe quedar adjunto a la factura.

ovnicraft replied:
in fact, creo mejor hacerlo con python pero la implementacion de xades tomara mas tiempo, o podemos usar plantillas para solo generar los hash del certificado, elemento, etc.
Mas que codigo debemos aportar con aclarar la especifcacion xades, y xmldsig-core del w3c

codeadict replied:
- Creo java es la única solución decente para Xades desgraciadamente, el xmlsec de C teoricamente lo soporta pero habría que probar está poco documentado y tomaría más tiempo.
- Se puede sacar primero una versión usable que funcione y después mejorarle
- el certificado puede estar en la BD aunque no sería lo más seguro.
- Pienso debería estar el RIDE en pdf que lo más complejo que veo es ese código de barras y junto el xml

ovnicraft replied:
Vamos por ahi entonces, ahora creo que todo lo que generemos podemos guardarlo en la base.
Queda un pendiente importante el path del certificado.
Creo que podemos poner el path como var en la compañia ? 
Sobre el RIDE, @alcidesrivera me comento que RML lo soporta, nos confirmas ? y una guia mucho mejor.

codeadict replied:
el RIDE solo tienen complejo el código de barras, que tipo de código sería, RIDE es algo específico del SRi o existe en inglés también esto??

codeadict replied:
Segun mi teoría se pudiera hacer en webkit mismo usando una fuente  como http://www.barcodesinc.com/free-barcode-font/ de ahí definimos un css como:

```
.ridebarcode {
    font-family: "Free 3 of 9";
    font-size: 36pt;
}
```

y luego en la plantilla webkit, por ejemplo:

```
<p class="ridebarcode">*${invoice.code}*</p>
```

así quedaría más limpio que usar RML :skull:

ovnicraft replied:
Gracias ! yo uso esto https://github.com/OCA/webkit-tools/tree/7.0/report_webkit_barcode y va de lujo

ovnicraft replied:
Solucionemos aca https://github.com/odoo-ecuador/odoo-ecuador/issues/51

codeadict replied:
![screenshot from 2015-01-15 00 22 06](https://cloud.githubusercontent.com/assets/168440/5752862/95d66d16-9c4c-11e4-91ef-f8548b859b7e.png)

codeadict replied:
ahh genial

2015-01-15 0:20 GMT-05:00 Cristian Salamea notifications@github.com:

> In l10n_ec_einvoice/xades/xades.py
> https://github.com/odoo-ecuador/odoo-ecuador/pull/48#discussion-diff-22991134
> :
> 
> > -            weight += 1
> > -            if weight > self._MODULO_11['MAX_WEIGHT']:
> > -                weight = self._MODULO_11['PESO']
> > -        mod = 11 - total % self._MODULO_11['BASE']
> >   +
> > -        mod = self._eval_mod11(mod)
> > -        return mod
> >   +
> >   +
> >   +class Xades(object):
> >   +
> > -    def apply_digital_signature(self, factura):
> > -        """
> > -        Metodo que aplica la firma digital al XML
> > -        """
> > -        OPT_PATH = '/opt/facturas/'
> 
> Gracias ! yo uso esto
> https://github.com/OCA/webkit-tools/tree/7.0/report_webkit_barcode y va
> de lujo
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/odoo-ecuador/odoo-ecuador/pull/48/files#r22991134.

## 

_Dairon Medina CaroOpenSource Consultantwww.gydsystems.com
http://www.gydsystems.com__email:_ info@gydsystems.com
_weblog:_ http://www.dairon.org
_mobile:_ _(+593)_-987612278
_LinkedIn:_ http://linkedin.com/in/codeadict
_GitHub:_ https://github.com/codeadict
_Skype:_ codeadict
_Twitter:_ @codeadict

codeadict commented on l10n_ec_einvoice/xades/xades.py:
```diff
@@ -0,0 +1,106 @@
+        # firma electrónica del xml
+        firma_path = os.path.join(os.path.dirname(__file__), JAR_PATH)
+        #
+        file_pk12 = ''
```
igual el pk12 deberia subirse en la configurración de la empresa pienso??

ovnicraft replied:
el path ? y la clave del certificado donde la ponemos ?

alcidesrivera commented:
Como claramente menciona el título es un WIP en donde se trabaja con directorios y valores quemados para irlos depurando. Todas las ideas son valiosas, aunque más lo son los aportes en código

Commit:
Descuento

Agregado campo para total de descuento por linea de factura y total
para la factura.

TEST TEST TEST

alcidesrivera commented:
Yo lo tengo hecho en el RML y me funciona de maravilla x q tenemos que utilizar tambien a parte del código de barras el logo de la compañia. Y así se reutiliza el account_print_invoice.rml propio del ERP

codeadict commented:
@alcidesrivera con el helper que muestra @ovnicraft podriamos migrar tu RML
a webkit para tener todos los reportes en un mismo formato

El 15 de enero de 2015, 0:23, alcidesrivera notifications@github.com
escribió:

> Yo lo tengo hecho en el RML y me funciona de maravilla x q tenemos que
> utilizar tambien a parte del código de barras el logo de la compañia
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/odoo-ecuador/odoo-ecuador/pull/48#issuecomment-70041342
> .

## 

_Dairon Medina CaroOpenSource Consultantwww.gydsystems.com
http://www.gydsystems.com__email:_ info@gydsystems.com
_weblog:_ http://www.dairon.org
_mobile:_ _(+593)_-987612278
_LinkedIn:_ http://linkedin.com/in/codeadict
_GitHub:_ https://github.com/codeadict
_Skype:_ codeadict
_Twitter:_ @codeadict

ovnicraft commented:
@alcidesrivera me sumo a @codeadict nos ayudas con el RML para verlo y ponerlo aca ? tenemos un issue abierto al respecto, un gist ayuda mucho alla.

codeadict commented on l10n_ec_einvoice/einvoice_view.xml:
```diff
@@ -0,0 +1,16 @@
+      <field name="inherit_id" ref="account.invoice_form"/>
+      <field name="arch" type="xml">
+	<button name="action_invoice_sent" position="before">
+	  <button name="action_generate_einvoice" type="object" string="Generar XML"/>
```
Por términos de UX mejor poner **Generar Factura Electrónica**, los mortales comunes no saben que es **XML**

Commit:
Correccion

Calculo de detalle de impuestos.
UX mejorado sugerencia @codeadict

ovnicraft commented:
@codeadict done !

alcidesrivera commented:
RIDE actualizado con los campos incrementados https://gist.github.com/alcidesrivera/772da5353a6cd1b321fb

codeadict commented:
que onda esto https://gist.github.com/alcidesrivera/aa27d891c996353a2b5e @alcidesrivera funciona?? de ser así no requerimos al maldito java de por medio

ovnicraft commented:
@codeadict deberia, pero hay que fix algunas cosas ! podemos hacerlo

alcidesrivera commented:
@codeadict @ovnicraft no funciono no genera correctamente los digest al
parecer. Tocaría seguirle revisando

El 15 de enero de 2015, 12:42 a. m., Dairon Medina C.<
notifications@github.com> escribió:

> que onda esto https://gist.github.com/alcidesrivera/aa27d891c996353a2b5e
> @alcidesrivera https://github.com/alcidesrivera funciona?? de ser así
> no requerimos al maldito java de por medio
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/odoo-ecuador/odoo-ecuador/pull/48#issuecomment-70042415
> .

alcidesrivera commented:
@ovnicraft @miltonlab @codeadict creo que a falta de los códigos ICE que seria muy largo hacerlo por el momento y la información adicional que dependera del giro de negocio de cada empresa podemos decir que tenemos un xml sin firmar bastante funcional, para así poder pasar a a la firma del xml

ovnicraft commented:
@alcidesrivera podemos entonces iniciar el desarrollo para la firma, con un diseño simple de backend con java y flexible para poder luego hacerlo con python.

Abri un issue en https://github.com/odoo-ecuador/odoo-ecuador/issues/52

codeadict commented:
Hola estoy desarrollando la firma en C con interface pra python pero me hace falta un xsd de la parte de firma, alguien sabe bien XSD que pueda ayudarme en eso, actualmente ya firmo con cualquier tipo de certificado pero lo que pone en el xml no es igual que el del sri, esta lib le voy a sacar open source

ovnicraft commented:
@codeadict genial ! puedes poner un PR ? 
Aca lo solicitado http://www.sri.gob.ec/web/guest/10116
